packages/playground-examples/copy/en/4-0/New TS Features/Variadic Tuples.ts

```diff
@@ -0,0 +1,54 @@
+// Variadic Tuples gives tuples the ability to handle the rest operator (...)
+// to pass types through type checker in a way that works like generics.
+
+// This is quite an advanced topic, so if you get lost do not worry too much.
+// It builds on example:generic-functions and example:tuples
+
+// To start off, here is a variadic tuple which will always prefix another
+// tuple with a number:
+
+type AddMax<T extends unknown[]> = [max: number,  ...rest: T];
+//          ^ Generic used to constrain the T
+//                                                ^ ... used to indicate where to merge
+
+// This can then be used for composition:
+type MaxMin = AddMax<[min: number]>
+type MaxMinDiameter = AddMax<[min: number, diameter: number]>
+
+// The same can be used after the tuple:
+type SuffixDIContext<T extends unknown[]> = [...first: T, context: any];
+type DIContainer = SuffixDIContext<[param: string]>
+
+// This mechanism can be combined with multiple input params. For example, This
+// function merges two arrays but uses '\0' as a sigil to indicate where the arrays 
+// start and stop.
+function joinWithNullTerminators<T extends unknown[], U extends unknown[]>(t: [...T], u: [...U]) {
+    return ['\0', ...t, '\0', ...u, '\0'] as const;
+}
+
+// Code flow analysis can infer the return type of a function like this
+// via contextual typing:
```

Comment:
```suggestion
// TypeScript can infer the return type of a function like this:
```

I don't think the references to control flow or contextual typing are correct here.